commit 28328ec87b45725f62bed1104d99c8b3220d1675
Author: Melanie Plageman <melanieplageman@gmail.com>
Date:   Wed Dec 18 18:16:43 2024 -0500

    Fix overflow danger in SampleHeapTupleVisible()
    
    68d9662be1c4b70 made HeapScanDesc->rs_ntuples unsigned but neglected to
    change how it was being used in SampleHeapTupleVisible().
    
    Return early if rs_ntuples is 0 to avoid overflowing and incorrectly
    executing the loop code in SampleHeapTupleVisible().
    
    Reported-by: Ranier Vilela
    Discussion: https://postgr.es/m/CAEudQAot_xQoZyPZjpj1aBUPrPykY5mOPHGyvfe%3Djz%2BWowdA3A%40mail.gmail.com

diff --git a/src/backend/access/heap/heapam_handler.c b/src/backend/access/heap/heapam_handler.c
index d0e5922eed7..689773ff239 100644
--- a/src/backend/access/heap/heapam_handler.c
+++ b/src/backend/access/heap/heapam_handler.c
@@ -2575,25 +2575,31 @@ SampleHeapTupleVisible(TableScanDesc scan, Buffer buffer,
 {
 	HeapScanDesc hscan = (HeapScanDesc) scan;
 
 	if (scan->rs_flags & SO_ALLOW_PAGEMODE)
 	{
+		uint32		start,
+					end;
+
+		if (hscan->rs_ntuples == 0)
+			return false;
+
 		/*
 		 * In pageatatime mode, heap_prepare_pagescan() already did visibility
 		 * checks, so just look at the info it left in rs_vistuples[].
 		 *
 		 * We use a binary search over the known-sorted array.  Note: we could
 		 * save some effort if we insisted that NextSampleTuple select tuples
 		 * in increasing order, but it's not clear that there would be enough
 		 * gain to justify the restriction.
 		 */
-		int			start = 0,
-					end = hscan->rs_ntuples - 1;
+		start = 0;
+		end = hscan->rs_ntuples - 1;
 
 		while (start <= end)
 		{
-			int			mid = (start + end) / 2;
+			uint32		mid = (start + end) / 2;
 			OffsetNumber curoffset = hscan->rs_vistuples[mid];
 
 			if (tupoffset == curoffset)
 				return true;
 			else if (tupoffset < curoffset)
